Use 'get a structured header' term from Fetch

zcorpan opened the pull request:
See whatwg/fetch#983

Change duplicate requirements in the Framework section to non-normative note.

Fail parsing if SH parameters are present, until SH is more stable.

zcorpan committed:
Use 'get a structured header' term from Fetch

See https://github.com/whatwg/fetch/pull/983

Change duplicate requirements in the Framework section to non-normative note.

Fail parsing if SH parameters are present, until SH is more stable.

zcorpan commented:
Why isn't https://tools.ietf.org/html/draft-ietf-httpbis-header-structure-14#section-4.2 used as the entry point with the header's byte sequence value? It seems the spec now doesn't skip OWS and doesn't support parameters. Is that intentional?

mikewest commented on index.bs:
```diff
@@ -58,8 +58,7 @@ urlPrefix: https://tools.ietf.org/html/draft-ietf-httpbis-header-structure; spec
         text: serialize Structured Header; url: #section-4.1
-        text: Structured Header parsing algorithm; url: #section-4.2.7
-urlPrefix: https://html.spec.whatwg.org/
+        text: Structured Header Token parsing algorithm; url: #section-4.2.6 urlPrefix: https://html.spec.whatwg.org/
```
1. The `urlPrefix` bit looks like things might have gotten a little mixed up...

2.  My understanding of Structured Headers is that the entry point for parsing ought to be the general parsing algorithm in 4.2 (https://tools.ietf.org/html/draft-ietf-httpbis-header-structure-14#section-4.2), not the specific types.

zcorpan replied:
1. Uh, I don't know how that `urlPrefix` got there... edit: oh, I accidentally removed a linebreak!
2. OK. Since the input to that is a byte sequence, we should remove the isomorphic decoding in step 1.

mikewest reviewed (changes requested):
Thanks!

zcorpan commented:
OK I've changed this to call the main entry point 4.2 with the byte sequence, and handle the returned tuple (bare_item, parameters).

It wasn't clear to me whether Structured Headers would fail parsing if the type isn't a token, so I explicitly said that should happen.

I also banned parameters, but maybe we want to ignore parameters instead?

I also removed an inline issue since [that issue](https://github.com/httpwg/http-extensions/issues/662) appears to be resolved, and Structured Headers how takes a byte sequence as input and converts it to an ASCII string.

mikewest commented:
@mnot and @annevk might have opinions about how they'd like this integration to work from a Structured Headers/Fetch standpoint. It's probably reasonable, for example, to add something like a "get a structured header `name` from `list` algorithm to header lists that would take care of the encoding questions?

In the status quo, this patch matches my understanding of what we need to do. SH gives you a raw parse of the header, and then it falls upon the spec to determine whether the result is acceptable.

mikewest commented on index.bs:
```diff
@@ -161,19 +162,16 @@ To <dfn abstract-op local-lt="parse header">obtain a response's embedder policy<
-    2.  Let |parsed policy| be the result of executing the [$Structured Header parsing algorithm$]
-        with <var ignore>input_string</var> set to |header|, and <var ignore>header_type</var> set
-        to "`item`".
+        If |parsed policy| is not a token, fail parsing.
```
```suggestion
        If |parsed policy| is not a [=structured header/token=], fail parsing.
```

annevk commented on index.bs:
```diff
@@ -144,7 +144,8 @@ The only currently valid `Cross-Origin-Embedder-Policy` value is "`require-corp`
 this header if it contains an invalid value. Likewise, user agents MUST ignore this header if the
-value cannot be parsed as a <a grammar>`sh-token`</a>.
+value cannot be parsed as a <a grammar>`sh-token`</a>. Parameters are not allowed; user agents MUST
+ignore this header if any parameters are present.
```
Are these requirements novel or do they duplicate the processing model? I kinda hope the latter in which case this should be a note.

mikewest replied:
They duplicate the processing model. Marking them as a note is fine; they aren't the normative description.

zcorpan replied:
I've changed MUSTs to will. That makes this a statement of fact. Do you think it should be a note?

annevk replied:
Seems fine either way, though I do worry a tiny bit about readers that aren't careful.

zcorpan replied:
Changed to a note.

annevk commented on index.bs:
```diff
@@ -161,19 +162,16 @@ To <dfn abstract-op local-lt="parse header">obtain a response's embedder policy<
 
-    1.  Set |header| to the result of [=isomorphic decoding=] |header|.
+    1.  Let |parsed policy| and |parsed parameters| be the result of executing
+        [$Parsing Header Fields into Structured Headers$] with <var ignore>input_string</var> set to
```
This algorithm returns a structured header value, no? Why do you expect two return values?

mikewest replied:
As currently written, https://tools.ietf.org/html/draft-ietf-httpbis-header-structure-14#section-4.2 returns the output of https://tools.ietf.org/html/draft-ietf-httpbis-header-structure-14#section-4.2.3, which returns "a (bare_item,
   parameters) tuple". I agree that this isn't entirely clear from the document.

annevk replied:
Well, I'd recommend treating it as a tuple in that case and not treating it as multiple return values. (Both of these things exist in other specifications as independent concepts.)

mikewest replied:
Along the lines of "Let |parsed| be the (bare_item, parameters) [=tuple=] resulting from executing ..."?

(It would be nice if we could define these in a way bikeshed could pick up. Perhaps we can add some sort of "Integration with Structured Headers" section to Fetch?)

annevk replied:
Yeah, as mentioned in my overall review comment I think we should add this to Fetch. In particular, add an algorithm that takes a header list, name, and an expected type? And perhaps the expected type is optional in case you support multiple different ones if that's a thing?

mikewest replied:
Ha. I missed the comment where you said all the things I've said. Sorry about that. I'll stop loudly agreeing with you and try to send you a patch instead.

mikewest replied:
https://github.com/whatwg/fetch/pull/983.

annevk reviewed (commented):
Yeah, I would prefer we put an abstraction in Fetch.

In https://github.com/whatwg/fetch/issues/930 I sketched something where for a header list there's an algorithm that takes a _name_ and a _type_ and returns failure or a representation of _type_. I think that still makes sense.

One other consideration I suppose is whether we should map structured header values to Infra values somehow for ease of use in specification algorithms and compatibility with IDL.

mikewest commented on index.bs:
```diff
@@ -161,19 +162,16 @@ To <dfn abstract-op local-lt="parse header">obtain a response's embedder policy<
-        passing it into the parsing algorithm?
-
-    3.  If |parsed policy| is "`require-corp`", set |policy| to "`require-corp`".
+    2.  If parsing did not fail and |parsed policy| is "`require-corp`", then set |policy| to
```
This is fine for the moment, but I don't really like the implicitness of "parsing failed". I'd prefer for the parsing algorithm to return something that we interpret as failure. Like `null`? (Obviously, this isn't the first place we're using this pattern, URL parsing comes to mind as well, but I don't like it there either. :) )

mikewest replied:
@annevk @mnot

annevk replied:
I don't really mind how Structured Headers words it, but I agree that whatever we add in Fetch will have to do something like that. Bridging with the IETF world is always going to look a little clunky I suppose, though this is miles ahead of the pack.

(I suspect we want null for header not present, failure for failures, and instance of typed thingie.)

mikewest replied:
I see now that URL parser does "return failure", which is nicer than I remember. That explicit model seems reasonable to me. It might be nice to define that pattern via Infra?

mnot replied:
@mikewest could you open an issue on SH regarding this?

mnot commented:
I probably do have opinions, but I'm out until the first week in Jan...

zcorpan commented:
In https://github.com/web-platform-tests/wpt/pull/20852#discussion_r359564886 @domenic said that parameters should be ignored, not fail parsing. This PR says to fail parsing. Which should we do?

From the Structured Headers spec, I think the intent is that unknown parameters should generally be ignored.

mikewest commented:
From structured headers' perspective, we should be able to parse headers with all the junk that can be attached to bits and pieces of the header's value according to the SH grammar. From this specific header's perspective, it may be reasonable to impose additional restrictions on the header's syntax.

My main concern here is forward compatibility with new things we introduce. In particular for COEP, I expect that we'll want to introduce something like `require-cors` in the future (because one-character differences that cause massive behavioral shifts are awesome!). I have no plans to introduce any parameters to `require-corp`, but who knows. The future is dark and murky. It might be reasonable to treat `require-corp;param1=value1` as `require-corp`, and it might be reasonable to ignore the header entirely.

It occurs to me now that we've dealt with unknown in the past for things like Referrer-Policy by ignoring the whole header, which allowed developers to do things like `Referrer-Policy: new-funky-value, old-boring-fallback;`. This model is a little more complicated in SH, because if we parse the header as a single item, it'll fail. We'd need to parse it as a list, and then ignore individual items in the list if they don't make sense, and then ensure that there aren't any additional items after the item we picked.

That's a bit weird.

I wonder if @mnot has opinions about this from the HTTP side (though I think he's going to be out until January), or @annevk from the Fetch side.

domenic commented:
The SH spec seems pretty clear on the intent:

> However, both Items and Inner Lists allow parameters as an extensibility mechanism; this means that values can later be extended to accommodate more information, if need be. As a result, header specifications are discouraged from defining the presence of an unrecognised parameter as an error condition.

I tend to agree with the SH spec, and don't see any reason to disallow parameters. They could be useful in the future and disallowing them just creates future compatibility problems, for no gain that I can see.

mikewest commented:
https://github.com/whatwg/fetch/commit/8ca61488c0c9efb32fd138ce14e25ce2b4ce0dfc landed in Fetch; @zcorpan, would you be interested in taking another pass at this, using that algorithm?

zcorpan commented:
Done.

This now has conflicts with master, but I'd like to squash before rebasing, since `index.html` is checked in I anticipate that will make a rebase without first squashing painful.

zcorpan commented:
I've squashed and rebased this now.

annevk commented on index.bs:
```diff
@@ -157,24 +157,22 @@ To <dfn abstract-op local-lt="parse header">obtain a response's embedder policy<
-        ISSUE(httpwg/http-extensions#662): The ASCII requirements of Structured Headers are
-        somewhat unclear to me. Do we need to check whether |header| is an [=ASCII string=] before
-        passing it into the parsing algorithm?
+    4. If |parameters| is not empty, then return |policy|.
```
I think we should ignore parameters as we do for `Content-Type`.

zcorpan replied:
Have you changed your opinion since https://github.com/whatwg/html/issues/5172#issuecomment-570288318 ?

I believe this PR is equivalent to "byte-case-sensitive comparison of the value with the allowed strings".

annevk replied:
Well, I think that's what we want long term. Perhaps instead we should add an open issue of sorts about changing that once SH is stable?

zcorpan replied:
I've filed https://github.com/whatwg/html/issues/5220

Do you think there should be an inline issue in the spec also?

annevk replied:
Yeah, I meant an inline marker.

zcorpan replied:
Done.

zcorpan committed:
Inline issue about handling of SH parameters

mikewest commented on index.bs:
```diff
@@ -157,24 +157,26 @@ To <dfn abstract-op local-lt="parse header">obtain a response's embedder policy<
-    3.  If |parsed policy| is "`require-corp`", set |policy| to "`require-corp`".
+        ISSUE: This step is expected to change when the Structured Headers for HTTP specification is
+        more stable. See [whatwg/html issue #5172](https://github.com/whatwg/html/issues/5172).
+        [[I-D.ietf-httpbis-header-structure]]
```
I'm fine with landing this just to correct the current text, but I think we're at the point where the IETF WG considers SH pretty close to done. I'd be surprised if something like parameterized items changed substantially at this point.

mikewest replied:
(So, LGTM from my side. If @annevk's also happy, land it. :))

annevk commented:
Did 31591be0f53480522d8930d397ade30e888b4b21 (#8) supersede this?

zcorpan commented:
Looks like it, @annevk. I'll close this. cc @yutakahirano
